x/pairing/keeper/msg_server_unfreeze.go

```diff
@@ -0,0 +1,39 @@
+package keeper
+
+import (
+	"context"
+	"strconv"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lavanet/lava/utils"
+	epochstoragetypes "github.com/lavanet/lava/x/epochstorage/types"
+	"github.com/lavanet/lava/x/pairing/types"
+)
+
+func (k msgServer) Unfreeze(goCtx context.Context, msg *types.MsgUnfreeze) (*types.MsgUnfreezeResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	providerAddr, err := sdk.AccAddressFromBech32(msg.GetCreator())
+	if err != nil {
+		return nil, utils.LavaFormatError("Unfreeze_get_provider_address", err, &map[string]string{"providerAddress": msg.GetCreator()})
+	}
+
+	for _, chainId := range msg.GetChainIds() {
+		stakeEntry, found, index := k.epochStorageKeeper.GetStakeEntryByAddressCurrent(ctx, epochstoragetypes.ProviderKey, chainId, providerAddr)
+		if !found {
+			continue
+		}
+
+		nextEpoch, err := k.epochStorageKeeper.GetNextEpoch(ctx, uint64(ctx.BlockHeight()))
+		if err != nil {
+			return nil, utils.LavaFormatError("Unfreeze_get_next_epoch", err, &map[string]string{"block": strconv.FormatInt(ctx.BlockHeight(), 10)})
+		}
+
+		// unfreeze the provider by making the StakeAppliedBlock be the next epoch. This will let the provider be added to the pairing list in the next epoch
+		stakeEntry.StakeAppliedBlock = nextEpoch
```

Comment:
Unfreeze should fail if the provider is not currently frozen - please add a check, and a test case.

Reply:
this is done in a way that we dont need to fail, unfreezing a normal provider will do nothing

Reply:
It's useful for the user to have feedback: if "unfreeze" succeeds, it means that the provider was frozen; if it fails, it mean that it wasn't (or that it wasn't staked for the respective chain-id).